### PR TITLE
Allow courses to control feature flags in dev mode via infoCourse.json

### PR DESCRIPTION
**Pull request by nwalters512**
This PR adds support for settings feature flags in `infoCourse.json`:

```json
{
  "options": {
    "devModeFeatures": ["question-sharing"]
  }
}
```

As the name implies, this option will only be honored in dev mode.

This functionality will be used in an upcoming change where we'll make R support in `server.py` toggleable by admins to avoid the proliferation of R-based questions, which in turn will let us work towards eliminating `rpy2`.

### CODE REVIEW

File: apps/prairielearn/src/lib/features/manager.ts [hunk 1]
```diff
@@ -98,11 +103,26 @@ export class FeatureManager<FeatureName extends string> {
...
-    if (featureIsEnabled) return true;
 
-    // Default to disabled if not explicitly enabled by a specific grant or config.
+    if (featureIsEnabled.has_feature_grant) return true;
+
+    // Allow features to be enabled in dev mode via `options.devModeFeatures`
+    // in `infoCourse.json`.
+    //
+    // We could check for the feature in the array in SQL, but it's easier to
+    // write this more defensively in JavaScript.
```
Comment (mwest1066) [inline apps/prairielearn/src/lib/features/manager.ts:115]: Definitely agreed that I prefer this in JS.

**Review by mwest1066 — APPROVED** [review]

**Comment by jonatanschroeder** [comment]
Would it make sense to add a sync warning if the feature is not enabled in production? To avoid confusion if the behavior locally does not match production.

**Comment by jonatanschroeder** [comment]
> Would it make sense to add a sync warning if the feature is not enabled in production? To avoid confusion if the behavior locally does not match production.

To clarify: I'm considering cases where an instructor might have a feature enabled in one course, then copy this setting to another course they teach (or send the setting to another instructor) assuming it will be enough to make it work, and it works locally but not in production. I believe having this be a warning in production means that an instructor will at least be aware of the reason why a particular setting is not working as expected in production.

Other than that: I like this setting, and I see how this can be used in other feature-flag cases besides rpy2.

**Comment by nwalters512** [comment]
@jonatanschroeder I like that idea, I implemented it in 0b43ae0. Mind giving that a look?

File: apps/prairielearn/src/sync/syncFromDisk.js [hunk 2]
```diff
@@ -39,12 +39,12 @@ const perf = makePerformance('sync');
  * @param {any} logger
  * @returns Promise<SyncResults>
  */
-export async function syncDiskToSqlWithLock(courseDir, courseId, logger) {
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/sync/syncFromDisk.js:42]: Nit: can't comment on three lines above, but the order in the comment doesn't match the arguments now.

**Review by jonatanschroeder — COMMENTED** [review]
I won't be able to test the code now, but it looks fine. I don't understand exactly why you changed the order of arguments, can you elaborate?

File: apps/prairielearn/src/sync/syncFromDisk.js (hunk 2 above)
Comment (nwalters512) [reply apps/prairielearn/src/sync/syncFromDisk.js:42]: Good catch, fixed in 0e94d34.

**Comment by nwalters512** [comment]
> I don't understand exactly why you changed the order of arguments, can you elaborate?

To me, it make more sense to put the identifier first 🤷

File: apps/prairielearn/src/sync/course-db.js [hunk 3]
```diff
@@ -758,6 +762,33 @@ export async function loadCourseInfo(coursePath) {
   const topics = getFieldWithoutDuplicates('topics', 'name');
   const assessmentModules = getFieldWithoutDuplicates('assessmentModules', 'name');
 
+  // In production environments, check if any features specified in `devModeFeatures`
+  // aren't also enabled for the course. If so,
```
Comment (mwest1066) [inline apps/prairielearn/src/sync/course-db.js:766]: Comment ends unexpectedly!

**Review by mwest1066 — APPROVED** [review]

File: apps/prairielearn/src/sync/course-db.js (hunk 3 above)
Comment (nwalters512) [reply apps/prairielearn/src/sync/course-db.js]: Good catch! I removed it entirely in a818818, I think it's redundant given comments on the following lines.

**Pull request merged**